### PR TITLE
catch when there are no changes in revert

### LINKED ISSUE #6917: lakectl: revert a non existing commit succeeds and creates a commit with no changes
Executing lakectl branch revert with a reference which does not exist on the branch results in a commit with no changes

image

image

### PR DESCRIPTION
**Pull request by lynnro314**
close #6917
Also added a flag to allow an empty revert

### CODE REVIEW

File: api/swagger.yml [hunk 1]
```diff
@@ -503,6 +503,10 @@ components:
         force:
           type: boolean
           default: false
+        allow_empty:
+          type: boolean
+          default: false
+          description: allow reverts without any changes
```
Comment (nopcoder) [inline api/swagger.yml:509]: allow empty commit (revert without changes)

File: cmd/lakectl/cmd/branch_revert.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 const (
 	branchRevertCmdArgs  = 2
 	ParentNumberFlagName = "parent-number"
+	allowEmptyRevert     = "allow-empty-revert"
```
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_revert.go:15]: allow-empty-commit?

There are two situations where we might want to allow empty commits: during regular commits and during reverts. I propose that we keep it the same.

What do you think?

File: cmd/lakectl/cmd/branch_revert.go [hunk 3]
```diff
@@ -58,6 +61,7 @@ func init() {
 	AssignAutoConfirmFlag(branchRevertCmd.Flags())
 
 	branchRevertCmd.Flags().IntP(ParentNumberFlagName, "m", 0, "the parent number (starting from 1) of the mainline. The revert will reverse the change relative to the specified parent.")
+	branchRevertCmd.Flags().Bool(allowEmptyRevert, false, "allow revert with no changes")
```
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_revert.go:64]: same here as part of the description where we instruct that we allow empty commit as part of the revert operation.

File: pkg/catalog/catalog.go [hunk 4]
```diff
@@ -148,6 +148,7 @@ type RevertParams struct {
 	Reference    string // the commit to revert
 	ParentNumber int    // if reverting a merge commit, the change will be reversed relative to this parent number (1-based).
 	Committer    string
+	AllowEmpty   bool // allow revert with no changes
```
Comment (nopcoder) [inline pkg/catalog/catalog.go:151]: also here: allow empty commit

File: pkg/graveler/graveler.go [hunk 5]
```diff
@@ -2457,6 +2457,9 @@ func (g *Graveler) Revert(ctx context.Context, repository *RepositoryRecord, bra
 		}
 		// merge from the parent to the top of the branch, with the given ref as the merge base:
 		metaRangeID, err := g.CommittedManager.Merge(ctx, repository.StorageNamespace, branchCommit.MetaRangeID, parentMetaRangeID, commitRecord.MetaRangeID, MergeStrategyNone)
+		if (metaRangeID == branchCommit.MetaRangeID) && !commitParams.AllowEmpty {
+			return nil, ErrEmptyRevert
+		}
```
Comment (nopcoder) [inline pkg/graveler/graveler.go:2462]: ErrNoChanges ?

**Review by nopcoder — CHANGES_REQUESTED** [review]

**Review by arielshaqed — APPROVED** [review]
This is great, thanks!

One worry is that we _really_ want to test this.

* If there **is** a unit test for revert, please add a test for "no changes". 
* If tests for revert are entirely missing, please open an issue to test it
* If we have no tests _and_ already have an issue, then we just need to prioritize it...

File: cmd/lakectl/cmd/branch_revert.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
 const (
 	branchRevertCmdArgs  = 2
 	ParentNumberFlagName = "parent-number"
+	allowEmptyRevert     = "allow-empty-commit"
```
Comment (nopcoder) [inline cmd/lakectl/cmd/branch_revert.go:15]: we have `allowEmptyCommit` you can use

File: pkg/graveler/graveler.go [hunk 7]
```diff
@@ -2551,6 +2551,9 @@ func (g *Graveler) Revert(ctx context.Context, repository *RepositoryRecord, bra
 		}
 		// merge from the parent to the top of the branch, with the given ref as the merge base:
 		metaRangeID, err := g.CommittedManager.Merge(ctx, repository.StorageNamespace, branchCommit.MetaRangeID, parentMetaRangeID, commitRecord.MetaRangeID, MergeStrategyNone)
+		if (metaRangeID == branchCommit.MetaRangeID) && !commitParams.AllowEmpty {
```
Comment (nopcoder) [inline pkg/graveler/graveler.go:2554]: Missed this one in previs PR - we can't check the return value before we verified the error first.

**Review by nopcoder — CHANGES_REQUESTED** [review]
Minor comments to fix. Also there is a comment ariel added I don't think it was addressed.

**Review by nopcoder — APPROVED** [review]
LGTM!

**Comment by lynnro314** [comment]
Checked the timeout error on docs:
`[TIMEOUT] [https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id](https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id) | Timeout`
The link seems fine, so merging assuming the timeout is temporary. 

Also issue for adding unit test for revert: #7445

**Pull request merged**